pkg/apiserver/profiling/profile.go

```diff
@@ -15,64 +15,169 @@ package profiling
 
 import (
 	"context"
+	"flag"
 	"fmt"
 	"io"
 	"io/ioutil"
 	"net/http"
 	"os"
-	"os/exec"
 	"strconv"
+	"time"
+
+	"github.com/goccy/go-graphviz"
+	"github.com/google/pprof/driver"
+	"github.com/google/pprof/profile"
+	"github.com/pkg/errors"
 
 	"github.com/pingcap-incubator/tidb-dashboard/pkg/apiserver/utils"
 )
 
-func profileAndWriteSVG(ctx context.Context, target *utils.RequestTargetNode, fileNameWithoutExt string, profileDurationSecs uint, httpClient *http.Client, tls bool) (string, error) {
-	url, err := getProfilingURL(target, profileDurationSecs, tls)
+type flagSet struct {
+	*flag.FlagSet
+	args []string
+}
+
+func fetchPprof(ctx context.Context, httpClient *http.Client, target *utils.RequestTargetNode, fileNameWithoutExt, format string, profileDurationSecs uint, tls bool) (string, error) {
+	tmpfile, err := ioutil.TempFile("", fileNameWithoutExt)
 	if err != nil {
 		return "", err
 	}
-	req, err := http.NewRequest(http.MethodGet, url, nil)
-	if err != nil {
+	defer tmpfile.Close()
+	tmpPath := fmt.Sprintf("%s.%s", tmpfile.Name(), format)
+	format = "-" + format
+	args := []string{
+		format,
+		// prevent printing stdout
+		"-output", "dummy",
+		"-seconds", strconv.Itoa(int(profileDurationSecs)),
+	}
+	address := fmt.Sprintf("%s:%d", target.IP, target.Port)
+	args = append(args, address)
+	f := &flagSet{
+		FlagSet: flag.NewFlagSet("pprof", flag.PanicOnError),
+		args:    args,
+	}
+	if err := driver.PProf(&driver.Options{
+		Fetch:   &fetcher{ctx: ctx, httpClient: httpClient, target: target, output: tmpPath, tls: tls},
+		Flagset: f,
+		Writer:  &oswriter{output: tmpPath},
+	}); err != nil {
 		return "", err
 	}
-	req = req.WithContext(ctx)
+	return tmpPath, nil
+}
+
+func (f *flagSet) StringList(o, d, c string) *[]*string {
+	return &[]*string{f.String(o, d, c)}
+}
+
+func (f *flagSet) ExtraUsage() string {
+	return ""
+}
+
+func (f *flagSet) Parse(usage func()) []string {
+	f.Usage = usage
+	_ = f.FlagSet.Parse(f.args)
+	return f.Args()
+}
+
+func (f *flagSet) AddExtraUsage(eu string) {}
+
+// oswriter implements the Writer interface using a regular file.
+type oswriter struct {
+	output string
+}
+
+func (o *oswriter) Open(name string) (io.WriteCloser, error) {
+	f, err := os.Create(o.output)
+	return f, err
+}
+
+type fetcher struct {
+	ctx        context.Context
+	httpClient *http.Client
+	target     *utils.RequestTargetNode
+	output     string
+	tls        bool
+}
+
+func (f *fetcher) Fetch(src string, duration, timeout time.Duration) (*profile.Profile, string, error) {
```

Comment:
do we still need this function?

Reply:
Yes, we need to use it to get DOT output